kms/keysource.go

```diff
@@ -0,0 +1,143 @@
+	if err != nil {
+		return err
+	}
+	key.EncryptedKey = base64.StdEncoding.EncodeToString(out.CiphertextBlob)
+	return nil
+}
+
+func (key *KMSMasterKey) EncryptIfNeeded(dataKey string) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+func (key *KMSMasterKey) Decrypt() (string, error) {
+	k, err := base64.StdEncoding.DecodeString(key.EncryptedKey)
+	if err != nil {
+		return "", fmt.Errorf("Error base64-decoding encrypted data key: %s", err)
+	}
+	sess, err := key.createSession()
+	if err != nil {
+		return "", fmt.Errorf("Error creating AWS session: %v", err)
+	}
+
+	service := kms.New(sess)
+	decrypted, err := service.Decrypt(&kms.DecryptInput{CiphertextBlob: k})
+	if err != nil {
+		return "", fmt.Errorf("Error decrypting key: %v", err)
+	}
+	return string(decrypted.Plaintext), nil
+}
+
+func (key *KMSMasterKey) NeedsRotation() bool {
+	return time.Since(key.CreationDate) > (time.Hour * 24 * 30 * 6)
+}
+
+func (key *KMSMasterKey) ToString() string {
+	return key.Arn
+}
+
+func NewKMSMasterKeyFromArn(arn string) KMSMasterKey {
+	k := KMSMasterKey{}
+	arn = strings.Replace(arn, " ", "", -1)
+	roleIndex := strings.Index(arn, "+arn:aws:iam::")
+	if roleIndex > 0 {
+		k.Arn = arn[:roleIndex]
+		k.Role = arn[roleIndex+1:]
+	} else {
+		k.Arn = arn
+	}
+	k.CreationDate = time.Now()
+	return k
+}
+
+func KMSMasterKeysFromArnString(arn string) []KMSMasterKey {
+	var keys []KMSMasterKey
+	if arn == "" {
+		return keys
+	}
+	for _, s := range strings.Split(arn, ",") {
+		keys = append(keys, NewKMSMasterKeyFromArn(s))
+	}
+	return keys
+}
+
+func (k KMSMasterKey) createStsSession(config aws.Config, sess *session.Session) (*session.Session, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	stsService := sts.New(sess)
+	name := "sops@" + hostname
+	out, err := stsService.AssumeRole(&sts.AssumeRoleInput{
+		RoleArn: &k.Role, RoleSessionName: &name})
+	if err != nil {
+		return nil, err
+	}
+	config.Credentials = credentials.NewStaticCredentials(*out.Credentials.AccessKeyId,
+		*out.Credentials.SecretAccessKey, *out.Credentials.SessionToken)
+	sess, err = session.NewSession(&config)
+	if err != nil {
+		return nil, err
+	}
+	return sess, nil
+}
+
+func (k KMSMasterKey) createSession() (*session.Session, error) {
+	re := regexp.MustCompile(`^arn:aws:kms:(.+):([0-9]+):key/(.+)$`)
+	matches := re.FindStringSubmatch(k.Arn)
+	if matches == nil {
+		return nil, fmt.Errorf("No valid ARN found in %s", k.Arn)
+	}
+	config := aws.Config{Region: aws.String(matches[1])}
+	sess, err := session.NewSession(&config)
+	if err != nil {
+		return nil, err
+	}
+	if k.Role != "" {
+		return k.createStsSession(config, sess)
+	}
+	return sess, nil
+}
+
+func (k KMSMasterKey) ToMap() map[string]string {
+	out := make(map[string]string)
+	out["arn"] = k.Arn
+	if k.Role != "" {
+		out["role"] = k.Role
+	}
+	out["created_at"] = k.CreationDate.Format("2006-01-02T15:04:05Z")
```

Comment:
I think a better way to write this would be `k.CreationDate.UTC().Format(time.RFC3339)`